src/sys/ptrace.rs

```diff
@@ -140,3 +149,124 @@ pub fn ptrace_setsiginfo(pid: Pid, sig: &siginfo_t) -> Result<()> {
         Err(e) => Err(e),
     }
 }
+
+#[cfg(target_arch = "x86_64")]
+// We're going to export it anyway
+// TODO add enum Register for x86
+#[allow(non_camel_case_types)]
+pub enum Register {
+    R15 = 0 * 8,
+    R14 = 1 * 8,
+    R13 = 2 * 8,
+    R12 = 3 * 8,
+    RBP = 4 * 8,
+    RBX = 5 * 8,
+    R11 = 6 * 8,
+    R10 = 7 * 8,
+    R9 = 8 * 8,
+    R8 = 9 * 8,
+    RAX = 10 * 8,
+    RCX = 11 * 8,
+    RDX = 12 * 8,
+    RSI = 13 * 8,
+    RDI = 14 * 8,
+    ORIG_RAX = 15 * 8,
+    RIP = 16 * 8,
+    CS = 17 * 8,
+    EFLAGS = 18 * 8,
+    RSP = 19 * 8,
+    SS = 20 * 8,
+    FS_BASE = 21 * 8,
+    GS_BASE = 22 * 8,
+    DS = 23 * 8,
+    ES = 24 * 8,
+    FS = 25 * 8,
+    GS = 26 * 8,
+}
+
+type Word = usize;
+
+/// Makes the `PTRACE_SYSCALL` request to ptrace
+pub fn syscall(pid: Pid) -> Result<()> {
+    unsafe {
+        ptrace(
+            ptrace::PTRACE_SYSCALL,
+            pid,
+            ptr::null_mut(),
+            ptr::null_mut(),
+        ).map(|_| ()) // ignore the useless return value
+    }
+}
+
+/// Makes the `PTRACE_PEEKUSER` request to ptrace
+#[cfg(target_arch = "x86_64")]
+pub fn peekuser(pid: Pid, reg: Register) -> Result<Word> {
+    let reg_arg = (reg as i32) as *mut c_void;
+    unsafe { ptrace(ptrace::PTRACE_PEEKUSER, pid, reg_arg, ptr::null_mut()).map(|r| r as Word) }
+}
+
+/// Makes the `PTRACE_POKEUSER` request to ptrace
+///
+/// # Safety
+/// When incorrectly used, may change the registers to bad values,
+/// causing e.g. memory being corrupted by a syscall, thus is marked unsafe
+#[cfg(target_arch = "x86_64")]
+pub unsafe fn pokeuser(pid: Pid, reg: Register, val: Word) -> Result<()> {
+    let reg_arg = (reg as u64) as *mut c_void;
+    ptrace(ptrace::PTRACE_POKEUSER, pid, reg_arg, val as *mut c_void).map(|_| ()) // ignore the useless return value
+}
+
+/// Sets the process as traceable with `PTRACE_TRACEME`
+pub fn traceme() -> Result<()> {
+    unsafe {
+        ptrace(
+            ptrace::PTRACE_TRACEME,
+            Pid::from_raw(0),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        ).map(|_| ()) // ignore the useless return value
+    }
+}
+
+/// Makes the `PTRACE_PEEKDATA` request to ptrace
+///
+/// # Safety
+/// This function allows to access arbitrary data in the traced process
+/// and may crash the inferior if used incorrectly and is thus marked `unsafe`.
+pub unsafe fn peekdata(pid: Pid, addr: usize) -> Result<Word> {
+    ptrace(
+        ptrace::PTRACE_PEEKDATA,
+        pid,
+        addr as *mut c_void,
+        ptr::null_mut(),
+    ).map(|r| r as Word)
+}
+
+/// Makes the `PTRACE_POKEDATA` request to ptrace
+///
+/// # Safety
+/// This function allows to access arbitrary data in the traced process
+/// and may crash the inferior or introduce race conditions if used
+/// incorrectly and is thus marked `unsafe`.
+pub unsafe fn pokedata(pid: Pid, addr: usize, val: Word) -> Result<()> {
```

Comment:
I'd like to see some tests for these if possible, even if they just return errors, if for no other reason than to test the ergonomics of our wrappers. For example here I don't know how you'd generate `addr` since it's supposed to be a raw pointer type.

Reply:
One simple test is to cancel a syscall by changing its syscall number to 0 during the enter stage, and then reverting it during the exit stage.